tools/render-test/render.h

```diff
@@ -106,6 +93,231 @@ enum class PrimitiveTopology
+        PixelShaderResource,
+        NonPixelShaderResource,
+        GenericRead,
+        CountOf,
+    };
+
+        /// Binding flags describe all of the ways a resource can be bound - and therefore used
+    struct BindFlag
+    {
+        enum Enum
+        {
+            VertexBuffer            = 0x001,
+            IndexBuffer             = 0x002, 
+            ConstantBuffer          = 0x004, 
+            StreamOutput            = 0x008, 
+            RenderTarget            = 0x010, 
+            DepthStencil            = 0x020, 
+            UnorderedAccess         = 0x040, 
+            PixelShaderResource     = 0x080, 
+            NonPixelShaderResource  = 0x100, 
+        };
+    };
+
+        /// Combinations describe how a resource can be accessed (typically by the host/cpu)
+    struct AccessFlag
+    {
+        enum Enum
+        {
+            Read = 0x1,
+            Write = 0x2
+        };
+    };
+
+        /// Get the type
+    SLANG_FORCE_INLINE Type getType() const { return m_type; }
+        /// True if it's a texture derived type
+    SLANG_FORCE_INLINE bool isTexture() const { return int(m_type) >= int(Type::Texture1D); }
+        /// True if it's a buffer derived type
+    SLANG_FORCE_INLINE bool isBuffer() const { return m_type == Type::Buffer; }
+
+        /// For a usage gives the required binding flags
+    static const BindFlag::Enum s_requiredBinding[int(Usage::CountOf)]; 
+
+    protected:
+    Resource(Type type):
+        m_type(type)
+    {}
+
+    Type m_type;
+};
+
+class BufferResource: public Resource
+{
+    public:
+    typedef Resource Parent;
+
+    struct Desc
+    {
+        void init(size_t sizeInBytesIn)
+        {
+            bindFlags = 0;
+            cpuAccessFlags = 0;
+            sizeInBytes = sizeInBytesIn;
+            elementSize = 0;
+        }
+            /// Set up default parameters based on usage
+        void setDefaults(Usage initialUsage);
+
+        int bindFlags;          ///< Combination of Resource::BindFlag or 0 (and will use initialUsage to set)
+        int cpuAccessFlags;        ///< Combination of Resource::AccessFlag 
+
+        size_t sizeInBytes;     ///< Total size in bytes 
+        int elementSize;        ///< Get the element stride. If > 0, this is a structured buffer
+    };
+
+        /// Get the buffer description
+    SLANG_FORCE_INLINE const Desc& getDesc() const { return m_desc; }
+
+        /// Ctor
+    BufferResource(const Desc& desc):
+        Parent(Type::Buffer),
+        m_desc(desc)
+    {
+    }
+
+    protected:
+    Desc m_desc;
+};
+
+class TextureResource: public Resource
+{
+    public:
+    typedef Resource Parent;
+
+    struct SampleDesc
+    {
+        void init()
+        {
+            numSamples = 1;
+            quality = 0;
+        }
+        int numSamples;                     ///< Number of samples per pixel
+        int quality;                        ///< The quality measure for the samples
+    };
+    
+    struct Size
```

Comment:
Definitely not an important point, but I've found that the terminology around "dimensions" and "size" can get really confusing just because everything is so overloaded. For my own part I tend to use terms as follows:

* "rank" to when referring to "dimensionality" as "number of dimensions"
  * As a small detail, I tend to use "shape" when referring to things that can't be simplified down to rank (e.g., a `TextureCube` technically has a rank of 2, but its *shape* is not the same as a 2D texture)
* "axis" when referring to a "dimension" to mean one of X, Y, Z
* "extent"/"extents" when referring to how big something is along one/all axes

So FWIW, I might argue for a type like this being called `Extents` instead of `Size`, and I might argue for the `calcMaxDimension` function being named to something like `calcMaxExtent` if it does what I think it does (the only alternative in this case would be that it returns the *axis* along which the *extent* is maximized, but I can't see why you'd need that...).

Anyway, definitely not important.

Reply:
Ok I know where you coming from. Whilst naming the type 'Size' it didn't seem great. I felt that it wasn't so terrible though - because it's a child class of TextureResource.